Reachability slice requires function bodies

tautschnig opened the pull request:
Reachability slicing relies on the CFG. The CFG, however, will not
contain edges from a function call to the next instruction when no body
is available for the function call. Therefore, reachability slicing
requires two steps:

- The model library needs to be applied. CBMC already did so,
goto-instrument now does with this commit.
- Remaining function calls without body need to be replaced by
nondet-return-value assignments.

Fixes: #6394

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Remove reachability_slicert::operator() definition to cpp file

This is a non-trivial implementation with external dependencies.
Upcoming changes will make it even more complex.

tautschnig committed:
Reachability slice requires function bodies

Reachability slicing relies on the CFG. The CFG, however, will not
contain edges from a function call to the next instruction when no body
is available for the function call. Therefore, reachability slicing
requires two steps:

- The model library needs to be applied. CBMC already did so,
goto-instrument now does with this commit.
- Remaining function calls without body need to be replaced by
nondet-return-value assignments.

Fixes: #6394

NlightNFotis reviewed (approved):
Hi Michael, confirmed it works against the code in #6394 .

Thanks for handling this!

chrisr-diffblue commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1067,8 +1067,11 @@ void goto_instrument_parse_optionst::instrument_goto_program()
+  if(
+    cmdline.isset("add-library") || cmdline.isset("mm") ||
+    cmdline.isset("reachability-slice") ||
+    cmdline.isset("reachability-slice-fb") ||
```
Once upon a future... it'd be lovely if we had proper feature flags, grouping of feature flags, etc.... rather than sprinkling hardcoded strings all over the place :-)

tautschnig committed:
remove_calls_no_body: generate status output

Use of this functionality will make symex warnings about functions
without body go away. To make up for this, at least status output should
be produced to ensure that this possible soundness issue of the
verification result can be caught.
